kvbc/test/v4blockchain/v4_blockchain_test.cpp

```diff
@@ -740,6 +843,770 @@ TEST_F(v4_kvbc, delete_last_reachable) {
+    auto versioned_val = std::get<categorization::VersionedValue>(*(values[i]));
+    ASSERT_EQ(versioned_keys[i].substr(key_start), (versioned_val.data).substr(val_start));
+  }
+
+  blockchain.multiGetLatest("immutable", immutable_keys, values);
+  key_start = std::string("immutable_key_").size();
+  val_start = std::string("immutable_value_").size();
+  for (size_t i = 0; i < immutable_keys.size(); ++i) {
+    ASSERT_TRUE(values[i].has_value());
+    auto immutable_val = std::get<categorization::ImmutableValue>(*(values[i]));
+    ASSERT_EQ(immutable_keys[i].substr(key_start), (immutable_val.data).substr(val_start));
+  }
+
+  merkle_keys.clear();
+  versioned_keys.clear();
+  immutable_keys.clear();
+
+  std::set<std::string> deleted_keys;
+  uint64_t new_max_block = max_block;
+  delete_blocks(new_max_block,
+                num_merkle_each,
+                num_versioned_each,
+                num_immutable_each,
+                [&deleted_keys](const std::string& key, uint64_t blkid, uint32_t kid, uint32_t kid_max) -> bool {
+                  std::mt19937 rgen;
+                  std::uniform_int_distribution<uint32_t> dist(2, std::numeric_limits<uint32_t>::max());
+                  if (((blkid + kid) * dist(rgen)) % 2 == 0) {
+                    deleted_keys.insert(key);
+                    return true;
+                  }
+                  return false;
+                });
+  ASSERT_EQ(blockchain.getGenesisBlockId(), 1);
+  ASSERT_EQ(blockchain.getLastBlockId(), new_max_block);
+  // Checking again
+  for (uint64_t blk = max_block + 1; blk <= new_max_block; ++blk) {
+    for (uint32_t kid = 1; kid <= num_merkle_each; ++kid) {
+      std::string key = "merkle_key_" + std::to_string(blk) + "_" + std::to_string(kid);
+      std::string val_str = "merkle_value_" + std::to_string(blk) + "_" + std::to_string(kid);
+      merkle_keys.push_back(key);
+      auto val = blockchain.getLatest("merkle", key);
+      if (deleted_keys.count(key)) {
+        ASSERT_FALSE(val.has_value());
+      } else {
+        ASSERT_TRUE(val.has_value());
+        auto merkle_val = std::get<categorization::MerkleValue>(*val);
+        ASSERT_EQ(merkle_val.block_id, new_max_block);
+        ASSERT_EQ(merkle_val.data, val_str);
+      }
+    }
+
+    for (uint32_t kid = 1; kid <= num_versioned_each; ++kid) {
+      std::string key = "versioned_key_" + std::to_string(blk) + "_" + std::to_string(kid);
+      std::string val_str = "versioned_value_" + std::to_string(blk) + "_" + std::to_string(kid);
+      versioned_keys.push_back(key);
+      auto val = blockchain.getLatest("versioned", key);
+      if (deleted_keys.count(key)) {
+        ASSERT_FALSE(val.has_value());
+      } else {
+        ASSERT_TRUE(val.has_value());
+        auto ver_val = std::get<categorization::VersionedValue>(*val);
+        ASSERT_EQ(ver_val.block_id, new_max_block);
+        ASSERT_EQ(ver_val.data, val_str + std::string(1, v4blockchain::detail::LatestKeys::STALE_ON_UPDATE[0]));
```

Comment:
why here you add ```std::string(1, v4blockchain::detail::LatestKeys::STALE_ON_UPDATE[0])``` ?

Reply:
For versioned I saw this in handleCategoryUpdates. Need to discuss this with you.

Reply:
I think that all sources of confusion are due to https://github.com/cloudnoize/concord-bft/pull/14
Once you rebase on our master, it will be clear

Reply:
Yes I got it, so updated the same, thanks @cloudnoize